_posts/2022-09-14-utilizing-state-machines-in-front-end-development.md

```diff
@@ -0,0 +1,543 @@
+---
+layout: post
+title: Utilizing state machines in front-end development
+author: jtiala
+excerpt: State machines and statecharts are powerful computational models, that have been used for decades in software development. As front-end developers, we might have forgotten them for a while, but in recent years, there has been newly founded hype around them. And rightfully so!
+tags:
+  - state machines
+  - statecharts
+  - xstate
+  - JavaScript
+---
+
+The state machine is a very old concept. In fact, it even precedes modern computers. Despite its respectable age, the state machine is a super useful computational model to have in your toolbox of modern web development. Recently, there has been a lot of buzz around state machines and especially a popular state machine library called [XState](https://xstate.js.org).
+
+In this blog post, I'll give you a brief 101 lecture on state machines and their extension, statecharts. After that, we'll take a look at some real-world examples of how to use state machines in a React web app context.
+
+# What are finite state machines?
+
+Let us start with some theory. If you have studied computer science, you might remember this stuff from class. And even if not, you still might have heard the term _Finite State Machine_ or _FSM_ for short.
+
+So, what are the FSMs? According to [Wikipedia](https://en.wikipedia.org/wiki/Finite-state_machine), an FSM is...
+
+> ... an abstract machine that can be in **exactly one** of a finite number of states at any given time. The FSM can **change from one state to another** in response to some inputs; the change from one state to another is called a transition. An FSM is defined by a list of its **states**, its **initial state**, and the **inputs that trigger each transition**.
+
+A lot of jargon, I know. To put it simply, state machines are abstractions that are used to describe the behavior of some system, which might be an algorithm in your web app, a physical device in real life, or the whole solar system.
+
+One of the nicest things about state machines is that they are convenient to present visually. For example, let's take a look at a state machine representing a simplified version of a traffic light.
+
+![A visual presentation of a state machine representing a simple traffic light](/img/utilizing-state-machines-in-front-end-development/SimpleTrafficLight.png)
+
+The machine can be in three different _states_: `green`, `yellow`, or `red`. It can be in only one of these states at one time. On the left, the little dot with an arrow pointing to the green state marks _the initial state_ of the machine. Meaning that the green state is activated, when the machine is executed.
```

Comment:
In most countries, there's a combined red-yellow-state, but in this simplified version, the machine has three...

Reply:
Added your sugggested text!